File: mcts.js
class Node {
    constructor(board, parent) {
        this.board = board;
        if (this.board.isWin() || this.board.isDraw()) {
            this.isTerminal = true;
        } else {
            this.isTerminal = false;
        }
        this.isFullyExpanded = this.isTerminal;
        this.parent = parent;
        this.visits = 0;
        this.score = 0;
        this.children = {};
    }
}

class MCTS {
    search(initialState, maxIterations = 1000) {
        this.root = new Node(initialState, null);

        for (let i = 0; i < maxIterations; i++) {
            let node = this.select(this.root);
            let score = this.rollout(node.board);
            this.backpropagate(node, score);
        }

        return this.getBestMove(this.root, 0);
    }

    select(node) {
        while (!node.isTerminal) {
            if (node.isFullyExpanded) {
                node = this.getBestMove(node, 2);
            } else {
                return this.expand(node);
            }
        }
        return node;
    }

    expand(node) {
        let states = node.board.legalMoves();
        for (let state of states) {
            if (!node.children[state.position.toString()]) {
                let newNode = new Node(state, node);
                node.children[state.position.toString()] = newNode;

                if (states.length === Object.keys(node.children).length) {
                    node.isFullyExpanded = true;
                }
                return newNode;
            }
        }
    }

    rollout(board) {
        while (!board.isWin()) {
            try {
                board = board.legalMoves()[Math.floor(Math.random() * board.legalMoves().length)];
            } catch {
                return 0;
            }
        }
        if (board.player2 === 'x') return 1;
        else if (board.player2 === 'o') return -1;
    }

    backpropagate(node, score) {
        while (node !== null) {
            node.visits += 1;
            node.score += score;
            node = node.parent;
        }
    }

    getBestMove(node, explorationConstant) {
        let bestScore = -Infinity;
        let bestMoves = [];

        for (let childNode of Object.values(node.children)) {
            let currentPlayer = childNode.board.player2 === 'x' ? 1 : -1;

            let exploration = explorationConstant * Math.sqrt(Math.log(node.visits / childNode.visits));
            let exploitation = currentPlayer * childNode.score / childNode.visits;
            let moveScore = exploitation + exploration;

            if (moveScore > bestScore) {
                bestScore = moveScore;
                bestMoves = [childNode];
            } else if (moveScore === bestScore) {
                bestMoves.push(childNode);
            }
        }
        return bestMoves[Math.floor(Math.random() * bestMoves.length)];
    }
}

File: mcts.py
import math
import random

class Node():
    def __init__(self, board, parent):
        self.board = board
        if self.board.is_win() or self.board.is_draw():
            self.is_terminal = True
        else:
            self.is_terminal = False
        self.is_fully_expanded = self.is_terminal
        self.parent = parent
        self.visits = 0
        self.score = 0
        self.children = {}

class MCTS():
    def search(self, initial_state, max_iterations=1000):
        self.root = Node(initial_state, None)

        for _ in range(max_iterations):
            node = self.select(self.root)
            score = self.rollout(node.board)
            self.backpropagate(node, score)

        return self.get_best_move(self.root, 0)
    
    def select(self, node):
        while not node.is_terminal:
            if node.is_fully_expanded:
                node = self.get_best_move(node, 2)
            else:
                return self.expand(node)
        return node
    
    def expand(self, node):
        states = node.board.legal_moves()
        for state in states:
            if str(state.position) not in node.children:
                new_node = Node(state, node)
                node.children[str(state.position)] = new_node

                if len(states) == len(node.children):
                    node.is_fully_expanded = True
                return new_node
    
    def rollout(self, board):
        while not board.is_win():
            try:
                board = random.choice(board.legal_moves())
            except:
                return 0
        if board.player_2 == 'x': 
            return 1
        return -1
                
    def backpropagate(self, node, score):
        while node is not None:
            node.visits += 1
            node.score += score
            node = node.parent
    
    def get_best_move(self, node, exploration_constant):
        best_score = float('-inf')
        best_moves = []

        for child_node in node.children.values():
            if child_node.board.player_2 == 'x': current_player = 1
            elif child_node.board.player_2 == 'o': current_player = -1

            exploration = exploration_constant * math.sqrt(math.log(node.visits / child_node.visits))   
            exploitation = current_player * child_node.score / child_node.visits                                  
            move_score = exploitation + exploration
            
            if move_score > best_score:
                best_score = move_score
                best_moves = [child_node]
            elif move_score == best_score:
                best_moves.append(child_node)
        return random.choice(best_moves)




























